physical/alibaba/alibaba.go

```diff
@@ -0,0 +1,228 @@
+	bucket     string
+	client     *oss.Client
+	logger     log.Logger
+	permitPool *physical.PermitPool
+}
+
+// NewAlibabaBackend constructs an OSS backend using a pre-existing
+// bucket. Credentials can be provided to the backend, sourced
+// from the environment.
+func NewAlibabaBackend(conf map[string]string, logger log.Logger) (physical.Backend, error) {
+	endpoint := os.Getenv("ALIBABA_OSS_ENDPOINT")
+	if endpoint == "" {
+		endpoint = conf["endpoint"]
+		if endpoint == "" {
+			return nil, fmt.Errorf("'endpoint' must be set")
+		}
+	}
+
+	bucket := os.Getenv("ALIBABA_OSS_BUCKET")
+	if bucket == "" {
+		bucket = conf["bucket"]
+		if bucket == "" {
+			return nil, fmt.Errorf("'bucket' must be set")
+		}
+	}
+
+	accessKeyID := os.Getenv("ALIBABA_ACCESS_KEY_ID")
+	if accessKeyID == "" {
+		accessKeyID = conf["access_key_id"]
+		if accessKeyID == "" {
+			return nil, fmt.Errorf("'access_key_id' must be set")
+		}
+	}
+
+	accessKeySecret := os.Getenv("ALIBABA_ACCESS_KEY_SECRET")
+	if accessKeySecret == "" {
+		accessKeySecret = conf["access_key_secret"]
+		if accessKeySecret == "" {
+			return nil, fmt.Errorf("'access_key_secret' must be set")
+		}
+	}
+
+	options := func(c *oss.Client) {
+		c.Config.Timeout = 30
+	}
+
+	client, err := oss.New(endpoint, accessKeyID, accessKeySecret, options)
+	if err != nil {
+		return nil, err
+	}
+
+	bucketObj, err := client.Bucket(bucket)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = bucketObj.ListObjects()
+	if err != nil {
+		return nil, errwrap.Wrapf(fmt.Sprintf("unable to access bucket %q at endpoint %q: {{err}}", bucket, endpoint), err)
+	}
+
+	maxParStr, ok := conf["max_parallel"]
+	var maxParInt int
+	if ok {
+		maxParInt, err = strconv.Atoi(maxParStr)
+		if err != nil {
+			return nil, errwrap.Wrapf("failed parsing max_parallel parameter: {{err}}", err)
+		}
+		if logger.IsDebug() {
+			logger.Debug("max_parallel set", "max_parallel", maxParInt)
+		}
+	}
+
+	a := &AlibabaBackend{
+		client:     client,
+		bucket:     bucket,
+		logger:     logger,
+		permitPool: physical.NewPermitPool(maxParInt),
+	}
+	return a, nil
+}
+
+// Put is used to insert or update an entry
+func (a *AlibabaBackend) Put(ctx context.Context, entry *physical.Entry) error {
+	defer metrics.MeasureSince([]string{"alibaba", "put"}, time.Now())
+
+	a.permitPool.Acquire()
+	defer a.permitPool.Release()
+
+	bucket, err := a.client.Bucket(a.bucket)
+	if err != nil {
+		return err
+	}
+
+	err = bucket.PutObject(entry.Key, bytes.NewReader(entry.Value))
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Get is used to fetch an entry
+func (a *AlibabaBackend) Get(ctx context.Context, key string) (*physical.Entry, error) {
+	defer metrics.MeasureSince([]string{"alibaba", "get"}, time.Now())
+
+	a.permitPool.Acquire()
+	defer a.permitPool.Release()
+
+	bucket, err := a.client.Bucket(a.bucket)
+	if err != nil {
+		return nil, err
+	}
+
+	object, err := bucket.GetObject(key)
+	if err != nil {
+		switch err.(type) {
```

Comment:
If you pulled out the type here, you wouldn't have to pull it out twice below and it would make the code slightly more efficient. Example here. https://tour.golang.org/methods/16